mod: Beginner Guide: rewrote the gamma correction section (#349)
IrisShaders/DocsPage: src/content/docs/current/Guides/Your First Shader/0_intro.mdx

## OpenGL and GLSL

> Minecraft is built using OpenGL, a high-level graphics API. OpenGL itself is not a piece of software; rather, it merely describes a specification for creating computer graphics that hardware developers are responsible for implementing. OpenGL uses GLSL (OpenGL Shading Language), which is the language Minecraft shader packs are coded in. GLSL is syntactically similar to C but introduces many extra features specifically designed for working with graphics.

Reviewer: ```suggestion
Minecraft is built using OpenGL, a high-level graphics API. OpenGL itself is not a piece of software; rather, it merely describes a specification for creating computer graphics that hardware developers are responsible for implementing. OpenGL uses GLSL (OpenGL Shading Language), which is the language Minecraft shaderpacks are developed in. GLSL is syntactically similar to C but introduces many extra features specifically designed for working with graphics.
```

> Old versions of Minecraft used an ancient version of OpenGL - OpenGL 2.0. Most older shader packs, for this reason, were written using GLSL version 120. However, since Minecraft 1.17, Minecraft was updated to use OpenGL 3.2, which targets version 150. However, shader loader mods like Iris and Optifine allow the use of any version of OpenGL/GLSL the user's hardware supports. Since OpenGL 3.3 requires 'DX10 Class' hardware, any computer released in the last decade should support it, so this is what we will be using. OpenGL 3.3 uses GLSL version 330.

Reviewer: ```suggestion
Old versions of Minecraft used an ancient version of OpenGL - OpenGL 2.0. Most older shader packs, for this reason, were written using GLSL version 120. In version 1.17, Minecraft was updated to use OpenGL 3.2, which targets version 150. However, shader loader mods like Iris and Optifine allow the use of any version of OpenGL/GLSL the user's hardware supports. Since OpenGL 3.3 requires 'DX10 Class' hardware, any computer released in the last decade should support it, so this is what we will be using. OpenGL 3.3 uses GLSL version 330.
```

For the next part it is important to note that since most of the game's development was done on OpenGL 2.0, the rendering pipeline the game uses does not take advantage of modern hardware features.

## How Shaders Work

> A "shader" is, by definition, any program that is executed on the GPU. For Minecraft, these shader programs (also know as passes) can be sorted into two distinct categories: composite passes, which execute for the entire screen, and gbuffers passes, which execute only for specific geometry. Each individual program is composed of several shader stages, two of which are required: a vertex shader, which executes once for each vertex of the geometry, and a fragment shader, which executes once for every pixel that covers the geometry. You can optionally include a compute, geometry, and/or tesselation stage for each pass.

Reviewer: I wouldn't call them composite passes, rather fullscreen passes, and probably put both that and gbuffers in quotes

Author: My language here was taken directly from the [Iris Docs](https://shaders.properties/current/reference/programs/overview/). Notably, there, they're called "composite-style" and "gbuffers-style" passes. I agree with both of your points, though.

:::tip[CPU vs GPU]

> Odds are, all your previous code has been executed on the CPU. Shaders are instead executed on the GPU, also know as a graphics card or video card. CPUs are great at performing complex sequential tasks, whereas GPUs are great are executing simple tasks in parallel. In fact, modern high-end graphics cards are capable of performing *trillions* of calculations per second! You can read more about the differences [here](https://www.intel.com/content/www/us/en/products/docs/processors/cpu-vs-gpu.html).

Reviewer: ```suggestion
Odds are, any code you've written in the past was executed on the CPU. Shaders are instead executed on the GPU, (also know as a graphics or video card). CPUs are great at performing complex sequential tasks, whereas GPUs are great at executing simple tasks in parallel. In fact, modern high-end graphics cards are capable of performing *trillions* of calculations per second! You can read more about the differences [here](https://www.intel.com/content/www/us/en/products/docs/processors/cpu-vs-gpu.html).
```
